src/Processors/QueryPlan/Streaming/ReplayStep.cpp

```diff
@@ -0,0 +1,44 @@
+#include <memory>
+#include <utility>
+#include <Processors/QueryPlan/Streaming/ReplayStep.h>
+#include <QueryPipeline/QueryPipelineBuilder.h>
+#include <Processors/Streaming/ReplayStepTransform.h>
+#include <IO/Operators.h>
+
+
+namespace DB
+{
+namespace Streaming
+{
+
+static ITransformingStep::Traits getTraits()
+{
+    return ITransformingStep::Traits
+    {
+        {
+            .preserves_distinct_columns = true,
+            .returns_single_stream = false,
+            .preserves_number_of_streams = true,
+            .preserves_sorting = true,
+        },
+        {
+            .preserves_number_of_rows = false,
+        }
+    };
+}
+
+ReplayStep::ReplayStep(const DataStream & input_stream_, float replay_speed_) 
+    : ITransformingStep(input_stream_, input_stream_.header, getTraits())
+    ,replay_speed(replay_speed_)
+{
+
+}
+
+void ReplayStep::transformPipeline(QueryPipelineBuilder & pipeline, const BuildQueryPipelineSettings &)
+{
+    auto transform = std::make_shared<ReplayStepTransform>(pipeline.getHeader(), replay_speed);
+    pipeline.addTransform(std::move(transform));
```

Comment:
there are two scenarios:
1) one input
2) multiple inputs (multi-shards)

Use `addSimpleTransform()` to add a transform for each input,